src/DotNet.Status.Web/Controllers/AzurePipelinesController.cs

```diff
@@ -0,0 +1,377 @@
+            IOptions<BuildMonitorOptions> options,
+            ILogger<AzurePipelinesController> logger)
+        {
+            _gitHubClientFactory = gitHubClientFactory;
+            _options = options;
+            _logger = logger;
+            _clientLazy = new Lazy<AzureDevOpsClient>(BuildAzureDevOpsClient);
+            _projectMapping = new Lazy<Task<Dictionary<string,string>>>(GetProjectMappingInternal);
+        }
+
+        private AzureDevOpsClient Client => _clientLazy.Value;
+
+        private AzureDevOpsClient BuildAzureDevOpsClient()
+        {
+            BuildMonitorOptions.AzurePipelinesOptions o = _options.Value.Monitor;
+            return new AzureDevOpsClient(o.BaseUrl, o.Organization, o.MaxParallelRequests, o.AccessToken);
+        }
+
+        private async Task<Dictionary<string, string>> GetProjectMappingInternal()
+        {
+            var projects = await Client.ListProjectsAsync();
+            return projects.ToDictionary(p => p.Id, p => p.Name);
+        }
+
+        private async Task<string> GetProjectNameAsync(string id)
+        {
+            Dictionary<string, string> mapping = await _projectMapping.Value;
+            return mapping.GetValueOrDefault(id);
+        }
+
+        [HttpPost]
+        [Route("build-complete")]
+        [ProducesResponseType((int) HttpStatusCode.NoContent)]
+        public async Task<IActionResult> BuildComplete(AzureDevOpsEvent<AzureDevOpsMinimalBuildResource> buildEvent)
+        {
+            _logger.LogInformation("Build complete notification for build '{buildId}', URL: {buildUrl}",
+                buildEvent.Resource.Id,
+                buildEvent.Resource.Url);
+            string projectName = await GetProjectNameAsync(buildEvent.ResourceContainers.Project.Id);
+            if (string.IsNullOrEmpty(projectName))
+            {
+                _logger.LogWarning(
+                    "Unknown project id '{projectId}' in collection '{collectionId}' reported in notification",
+                    buildEvent.ResourceContainers.Project.Id,
+                    buildEvent.ResourceContainers.Collection.Id
+                );
+                return NoContent();
+            }
+
+            _logger.LogInformation("Resolved build '{buildId}' for project '{project}, fetching build details",
+                buildEvent.Resource.Id,
+                projectName);
+            Build build = await Client.GetBuildAsync(projectName, buildEvent.Resource.Id);
+
+            if (IsIgnoredReason(build) ||
+                IsIgnoredStatus(build))
+            {
+                return NoContent();
+            }
+
+            await ProcessBuildNotificationsAsync(build);
+
+            return NoContent();
+        }
+
+        private bool IsIgnoredReason(Build build)
+        {
+            switch (build.Reason)
+            {
+                case "batchedCI":
+                case "buildCompletion":
+                case "individualCI":
+                case "manual":
```

Comment:
Shouldn't we ignore the manual ones? I mean, if someone manually queued it then they'll probably keep an eye on it. Right?

Reply:
We frequently re-queue builds, especially deployment ones.  Just because I click the button doesn't mean I'm paying attention to it (especially in things like OSOB, where a build can take 8ish hours).  And it's still a master build that has a failure we should be concerned with .